conn2res/reservoir.py

```diff
@@ -9,8 +9,223 @@
+            ctx = load_file('cortical.npy')
+            ctx = ctx[self.idx_node]
+
+            if node_set == 'ctx':
+                # select all nodes in cortex we want to select from
+                selected_nodes = np.where(ctx[nodes_from] == 1)[0]
+            elif node_set == 'subctx':
+                # select all nodes in subcortex we want to select from
+                selected_nodes = np.where(ctx[nodes_from] == 0)[0]
+
+            # remove nodes we do not want to select from
+            selected_nodes = np.setdiff1d(selected_nodes, nodes_without)
+
+        elif node_set == 'random':
+            # nodes we want to select from
+            nodes_from = np.setdiff1d(nodes_from, nodes_without)
+
+            # select random nodes
+            selected_nodes = np.random.choice(nodes_from, size=n_nodes)
+
+        elif node_set == 'shortest_path':
+            # calculate shortest paths between all nodes
+            D = distance_bin(self.w)
+            D = np.triu(D)  # remove repetitions
+
+            # nodes we want to select from
+            nodes_from = np.setdiff1d(nodes_from, nodes_without)
+
+            # shortest paths between all nodes of interest
+            D = D[np.ix_(nodes_from, nodes_from)]
+
+            # select all node pairs with requested shortest path from each other
+            if isinstance(kwargs['shortest_path'], str):
+                node_pairs = np.argwhere(D == np.amax(D))
+            elif isinstance(kwargs['shortest_path'], int):
+                node_pairs = np.argwhere(D == kwargs['shortest_path'])
+
+            # select requested number of nodes from the set above
+            if len(np.unique(node_pairs)) >= n_nodes:
+                i = 1
+                while len(np.unique(node_pairs[:i, :])) < n_nodes:
+                    i += 1
+                selected_nodes = nodes_from[np.unique(node_pairs[:i, :])]
+            else:
+                raise ValueError(
+                    'n_nodes do not exist with given shortest_path')
+
+        else:
+            # nodes we want to select from
+            nodes_from = np.setdiff1d(nodes_from, nodes_without)
+
+            # load resting-state networks and filter to active nodes
+            rsn_mapping = load_file('rsn_mapping.npy')
+            rsn_mapping = rsn_mapping[self.idx_node]
+
+            # get modules
+            module_ids, modules = get_modules(rsn_mapping)
+
+            if node_set in module_ids:
+                # select all nodes in the requested module
+                selected_nodes = [e for i, e in enumerate(
+                    modules) if (module_ids == node_set)[i]][0]
+
+                # intersection of nodes we want to select from
+                selected_nodes = np.intersect1d(selected_nodes, nodes_from)
+            else:
+                raise ValueError('node_set does not exist with given value')
+
+        return selected_nodes
+
+    def _update_component(self):
```

Comment:
This method makes sure that the connectivity matrix is fully connected, i.e. all nodes belong to the same component. It is essential after e.g., restricting the connectivity matrix to cortical nodes, as some nodes will drop out as they were connected only to subcortical regions.

This method is quite basic at the moment and I believe it works when there is only 1 large component with >1 nodes. It could be easily made more generic to work in cases where there are multiple large components by, for instance, choosing the largest one.